### ISSUE TITLE
[css-color-4] Spec doesn't specify whether implementation must honor <decimal> in rgb/rgba or if it can truncate to an <integer>

### ISSUE THREAD

**Issue by hichris1234** [w3c/csswg-drafts#1983]
[The RGB functions: rgb() and rgba()](https://drafts.csswg.org/css-color/#rgb-functions) states that:

> These values come from the fact that many graphics engines store the color channels internally as a single byte, which can hold integers between 0 and 255. However, the CSS syntax allows full \<number\>s, not just \<integer\>s, for authoring convenience.

However, this section of the spec is unclear as to whether an implementation must try to honor the color channels with decimals preserved or if the implementation can truncate/round the \<number\> to an integer. Edge has implemented this part of the spec, and appears to simply truncate the number. [Chromium is thinking about implementing it](https://groups.google.com/a/chromium.org/forum/#!topic/blink-dev/zH51Atd-Vag), but wants to clarify this section first.

**Comment by svgeesus** [comment]
Thanks for opening this, agree it needs to be clarified. The "for authoring convenience" is misleading and should be removed. This isn't just syntactic sugar.

Especially given the move to 10-bit and 12-bit per channel, the intent of the  the reason is to allow greater precision. Whether implementations `should` or `must` store the greater precision needs to be discussed. This also affects serialization.

**Comment by tabatkins** [comment]
We make no guarantees whatsoever on numeric precision - UAs can store data at whatever precision they want to store it at.  They could do 4-bit color if they felt like it, it would just be ugly on anything more modern than a console.

**Comment by annevk** [comment]
Rounding seems reasonable, but serializing an integer sometimes and sometimes something containing a dot seems like asking for trouble.

**Comment by annevk** [comment]
The other problem here is that if two UAs both use say 8-bit it's unclear whether 254.5 means 254 or 255. I think if UAs use an equal amount of bits they need to have the same behavior.

**Comment by Nadya678** [comment]
Is anyone able to see this difference?

**Comment by svgeesus** [comment]
> Is anyone able to see this difference? 

As we move to wider gamuts and higher dynamic range then yes, banding due to rounding to 8 bits is visibly noticeable. For sRGB, not really.

**Comment by tabatkins** [comment]
Yeah, specifying the required rounding behavior would probably be good.

**Comment by annevk** [comment]
We should also have tests ensuring the serialization of CSS colors used on `<canvas>` are not affected. At least not until `<canvas>` supports more than 8 bits.

And also, given that we require 8 bits there, I think we should require that as a minimum in CSS too for a browsers conformance class.

**Comment by tabatkins** [comment]
Yeah, for a browser conformance class 8 bits is a reasonable minimum.

(@fantasai and I tried to write up a comprehensive set of required minimums for CSS some time ago, but there was still too much variation between browsers. In particular, Opera did a lot of trickery with shoving multiple properties into the same bit field, so they had less bits available for values than you'd naively expect. That might be more plausible now.)

**Comment by csnardi** [comment]
Since there are some errors in the WPT rgb tests for decimals (fixed in https://github.com/w3c/web-platform-tests/pull/8538), I didn't realize Firefox has also implemented decimal channel parsing. As a result, I decided to do a comparison of Edge and Firefox. While Edge truncates decimals, Firefox rounds them (so in Edge `221.5` goes to `221` whereas in Firefox it would go to `222`).

It seems important to clarify this issue. Rounding would likely be more "accurate" than truncating, since it would always create a value closer to what was originally intended when compared to truncating.

**Comment by svgeesus** [comment]
> Rounding would likely be more "accurate" than truncating, since it would always create a value closer to what was originally intended when compared to truncating.

Yes, clearly rounding is correct here

**Comment by csnardi** [comment]
I opened #2086 for a potential way to clarify the wording in this section. This should have the side effect of clarifying how any values eventually resolved to RGB/RGBA should be computed (e.g. HSL/HSLA).
